### PR TITLE
refactor: make MenuItem use our own props

### PR DESCRIPTION
**Pull request by jordankoschei-okta**
Refactor the internal `MenuItem` component to use our own props, rather than extending `MuiMenuItemProps`. Also removes the unnecessary `forwardRef`.

### CODE REVIEW

File: packages/odyssey-react-mui/src/MenuItem.tsx [hunk 1]
```diff
@@ -10,38 +10,33 @@
...
-  /**
-   * Toggles whether or not the MenuItem represents a destructive action.
-   */
+export type MenuItemProps = {
+  hasInitialFocus?: boolean;
+  children: ReactNode;
+  isSelected?: boolean;
   isDestructive?: boolean;
-}
+};
```
Comment (KevinGhadyani-Okta) [inline packages/odyssey-react-mui/src/MenuItem.tsx:22]: I think we should add `onClick`.

But also, it might be good to search Atko and see how this is currently used:
```tsx
<MenuItem content={undefined} rel={undefined} rev={undefined}>
<MenuItem value={DEFAULT_OPTION_VALUE} data-testid="select-option">
<MenuItem key={index} onClick={() => navigateTo(link)}>
<MenuItem value="true">
<MenuItem onClick={logOut}>{t('topBar.dropdown.signOut')}</MenuItem>
```

There are all the variations I found from a quick search.

Comment (jordankoschei-okta) [reply packages/odyssey-react-mui/src/MenuItem.tsx]: I added the following props:

- `content`
- `data-testid`
- `href`
- `onClick`
- `rel`
- `rev`
- `value`

I don't know what `rev` is either, so maybe we don't need that one.

For `onClick`, I have an optional `onClick` function and the existing `closeMenu` behavior you'd previously added.

Comment (francoislehoux-okta) [reply packages/odyssey-react-mui/src/MenuItem.tsx]: First one has these as undefined as current odyssey version breaks typescript, as reported a little while back :)

In reality, it would only be using `children`.

File: packages/odyssey-react-mui/src/MenuItem.tsx [hunk 2]
```diff
@@ -10,36 +10,48 @@
...
-    | "divider"
-    | "focusVisibleClassName"
-  > {
-  /**
-   * Toggles whether or not the MenuItem represents a destructive action.
-   */
+export type MenuItemProps = {
+  children: ReactNode;
+  content?: string;
+  "data-testid"?: string;
```
Comment (KevinGhadyani-Okta) [inline packages/odyssey-react-mui/src/MenuItem.tsx:25]: We can remove this prop.

File: packages/odyssey-react-mui/src/MenuItem.tsx [hunk 3]
```diff
@@ -10,36 +10,48 @@
...
+}: MenuItemProps) => {
   const { closeMenu } = useContext(MenuContext);
-  console.log(closeMenu);
+
+  const onClick = useCallback(
+    (event: React.MouseEvent<HTMLLIElement, MouseEvent>) => {
+      onClickProp?.(event);
+      closeMenu();
+    },
+    [onClickProp]
```
Comment (KevinGhadyani-Okta) [inline packages/odyssey-react-mui/src/MenuItem.tsx:48]: We need to add `closeMenu` in this array.

File: packages/odyssey-react-mui/src/MenuItem.tsx [hunk 4]
```diff
@@ -10,36 +10,48 @@
...
+  rev?: string;
+  value?: string;
+};
 
-const MenuItem = ({ isDestructive, ...props }: MenuItemProps) => {
+const MenuItem = ({
+  isDestructive,
+  onClick: onClickProp,
+  ...props
+}: MenuItemProps) => {
```
Comment (KevinGhadyani-Okta) [inline packages/odyssey-react-mui/src/MenuItem.tsx:40]: If we're defining `MenuItemProps`, we should be defining every prop here too, so it's clear to any dev what `...props` is. It also prevents people from bypassing TypeScript with an `@ts-ignore`.

File: packages/odyssey-react-mui/src/MenuItem.tsx [hunk 5]
```diff
@@ -10,40 +10,63 @@
...
   const { closeMenu } = useContext(MenuContext);
 
+  const onClick = useCallback(
+    (event: React.MouseEvent<HTMLLIElement, MouseEvent>) => {
+      onClickProp?.(event);
+      closeMenu();
+    },
+    [onClickProp, closeMenu]
+  );
```
Comment (KevinGhadyani-Okta) [inline packages/odyssey-react-mui/src/MenuItem.tsx:53]:
```ts
import { memo, useCallback, useContext, type MouseEventHandler, type ReactNode } from "react";
```

```suggestion
  const onClick = useCallback<MouseEventHandler<HTMLLIElement>>(
    (event) => {
      onClickProp?.(event);
      closeMenu();
    },
    [onClickProp, closeMenu]
  );
```

File: packages/odyssey-react-mui/src/MenuItem.tsx [hunk 6]
```diff
@@ -10,40 +10,63 @@
...
+
   return (
     <MuiMenuItem
-      {...props}
-      onClick={closeMenu}
+      content={content}
+      /* eslint-disable-next-line jsx-a11y/no-autofocus */
+      autoFocus={hasInitialFocus}
```
Comment (KevinGhadyani-Okta) [inline packages/odyssey-react-mui/src/MenuItem.tsx:59]: I'm really wondering if we should just remove the ESLint rule altogether. No one's gonna be randomly adding this prop to Odyssey components where it doesn't belong.

Thoughts?

File: packages/odyssey-react-mui/src/MenuItem.tsx [hunk 7]
```diff
@@ -10,40 +10,63 @@
...
+  content?: string;
+  hasInitialFocus?: boolean;
+  isSelected?: boolean;
   isDestructive?: boolean;
-}
+  onClick?: MuiMenuItemProps["onClick"];
+  rel?: string;
+  rev?: string;
+  value?: string;
+};
```
Comment (KevinGhadyani-Okta) [inline packages/odyssey-react-mui/src/MenuItem.tsx:32]: Can we add docs for these props? I specifically wanna know the `rel` and `rev` ones as they're not clear.

`isDestructive` is also unclear. Does that just turn it red? it doesn't change any functionality? Because at that point I'd consider it a `"variant"` of the component rather than a state.

Comment (KevinGhadyani-Okta) [reply packages/odyssey-react-mui/src/MenuItem.tsx:32]: `isDestructive` is also unclear. Does that just turn it red? it doesn't change any functionality? Because at that point I'd consider it a `"variant"` of the component rather than a state.

Comment (KevinGhadyani-Okta) [reply packages/odyssey-react-mui/src/MenuItem.tsx:32]: What's the difference between `children` and `content`? I know it's a MUI thing, but I can't tell from looking at the props.

File: packages/odyssey-react-mui/src/MenuItem.tsx (hunk 6 above)
Comment (jordankoschei-okta) [reply packages/odyssey-react-mui/src/MenuItem.tsx:59]: I'm in favor of removing it; the only time I'm even aware of it is when we're disabling it.

**Comment by jordankoschei-okta** [comment]
Great suggestions @KevinGhadyani-Okta!

File: packages/odyssey-react-mui/src/MenuItem.tsx (hunk 7 above)
Comment (jordankoschei-okta) [reply packages/odyssey-react-mui/src/MenuItem.tsx]: - I removed `rel` because it only applies if it's a link with an `href`, which `MenuItem` should never be
- I removed `rev` because it hasn't been an acceptable attribute since HTML 4 (yikes)
- I removed `content` because I have no idea what it does, can't find documentation in MUI, and can't find any legitimate use cases
- I changed `isDestructive` to be a `destructive` variant

File: packages/odyssey-react-mui/src/MenuItem.tsx [hunk 8]
```diff
@@ -10,40 +10,63 @@
...
+export type MenuItemProps = {
+  children: ReactNode;
+  hasInitialFocus?: boolean;
+  isSelected?: boolean;
   isDestructive?: boolean;
-}
+  onClick?: MuiMenuItemProps["onClick"];
+  value?: string;
+  variant?: "default" | "destructive";
+};
```
Comment (KevinGhadyani-Okta) [inline packages/odyssey-react-mui/src/MenuItem.tsx:36]: Can we document these or do you wanna do that in another PR?

I'm fine moving it to another PR, we just need a ticket for it.

File: packages/odyssey-react-mui/src/MenuItem.tsx [hunk 9]
```diff
@@ -10,40 +10,63 @@
...
+  children,
+  hasInitialFocus,
+  isSelected,
+  onClick: onClickProp,
+  value,
+  variant = "default",
+}: MenuItemProps) => {
   const { closeMenu } = useContext(MenuContext);
 
+  const onClick = useCallback<MouseEventHandler<HTMLLIElement>>(
```
Comment (KevinGhadyani-Okta) [inline packages/odyssey-react-mui/src/MenuItem.tsx:48]: This can actually be changed up to be relative to MUI's prop because that's where it's ultimately going to go:
```suggestion
  const onClick = useCallback<MuiMenuItemProps["onClick"]>(
```

File: packages/odyssey-react-mui/src/MenuItem.tsx (hunk 6 above)
Comment (KevinGhadyani-Okta) [reply packages/odyssey-react-mui/src/MenuItem.tsx:59]: Let's add a ticket to remove it from our ESLint rules and also remove the ESLint-disabling comments as well.

**Review by KevinGhadyani-Okta — CHANGES_REQUESTED** [review]
Requesting changes.

File: packages/odyssey-react-mui/src/MenuItem.tsx (hunk 9 above)
Comment (jordankoschei-okta) [reply packages/odyssey-react-mui/src/MenuItem.tsx:48]: Screenshot 2023-07-07 at 10 53 21 AM

It's not happy with me when I do this 🤔

Comment (jordankoschei-okta) [reply packages/odyssey-react-mui/src/MenuItem.tsx]: `NonNullable<MuiMenuItemProps["onClick"]>`
`<MuiMenuItemProps["onClick"] & {}>`

**Review by KevinGhadyani-Okta — APPROVED** [review]
Approved, but make that change noted.
